src/Microsoft.FeatureManagement/FeatureManager.cs

```diff
@@ -371,16 +368,53 @@ private async ValueTask<EvaluationEvent> EvaluateFeature<TContext>(string featur
                     await sessionManager.SetAsync(evaluationEvent.FeatureDefinition.Name, evaluationEvent.Enabled).ConfigureAwait(false);
                 }
 
-                if (evaluationEvent.FeatureDefinition.Telemetry != null &&
-                    evaluationEvent.FeatureDefinition.Telemetry.Enabled)
+                // Only add an activity event if telemetry is enabled for the feature
+                if (telemetryEnabled)
                 {
-                    PublishTelemetry(evaluationEvent, cancellationToken);
+                    AddEvaluationActivityEvent(evaluationEvent);
                 }
             }
 
             return evaluationEvent;
         }
 
+        private void AddEvaluationActivityEvent(EvaluationEvent evaluationEvent)
+        {
+            // Break if there is no activity, this shouldn't happen
+            if (Activity.Current == null)
+            {
+                return;
+            }
+
+            ActivityTagsCollection tags = new ActivityTagsCollection
+            {
+                { "TargetingId", evaluationEvent.TargetingContext?.UserId },
+                { "FeatureName", evaluationEvent.FeatureDefinition.Name },
+                { "Variant", evaluationEvent.Variant?.Name },
+                { "Enabled", evaluationEvent.Enabled },
+                { "VariantAssignmentReason", evaluationEvent.VariantAssignmentReason },
+                { "Version", ActivitySource.Version }
+            };
+
+            if (evaluationEvent.FeatureDefinition.Telemetry.Metadata != null)
+            {
+                foreach (KeyValuePair<string, string> kvp in evaluationEvent.FeatureDefinition.Telemetry.Metadata)
+                {
+                    try
+                    {
+                        tags.Add(kvp.Key, kvp.Value);
+                    } catch (InvalidOperationException)
+                    {
+                        Logger?.LogWarning($"{kvp.Key} from TelemetryMetadata will be ignored, as it would override an existing key.");
+                    }
+                }
+            }
+
+            ActivityEvent activityEvent = new ActivityEvent("feature_flag", DateTimeOffset.Now, tags);
```

Comment:
Consider evaluating the `DateTimeOffset.Now` closer to the actual flag check and passing it as an argument so it reflects the actual moment a bit more accurately. By evaluating here, any added delay from the extra processing would end up influencing the recorded time.

Maybe grab it even before the various `ISessionManager` calls (I don't know exactly what that abstraction is responsible for so take with a grain of salt).

Reply:
That would be more accurate for when the actual check started, but I think we'd rather be as close as possible to the return of the final result. Reason being so metrics emitted at or after this timestamp can be expected to use the new result. If we did it at the beginning and their implementation of `ISessionManager` was slow, the timestamps would be less reliable to correlate.

Reply:
Although- I could use the end of the activity to get the timestamp I'm referring to, so not sure if that would be as reliable though. But in that same line of thinking, the start of the activity could be the considered the start.